File: apps/tenant/src/types/pos.ts
/**
 * POS TypeScript Types
 * Definiciones de tipos para el módulo POS
 */

export interface POSRegister {
  id: string
  tenant_id: string
  store_id?: string
  name: string
  active: boolean
  created_at: string
}

export interface POSShift {
  id: string
  register_id: string
  opened_by: string
  opened_at: string
  closed_at?: string
  opening_float: number
  closing_total?: number
  status: 'open' | 'closed'
  difference?: number
  total_sales?: number
  cash_sales?: number
  expected_cash?: number
  counted_cash?: number
  loss_amount?: number
  loss_note?: string | null
}

export interface POSReceiptLine {
  id?: string
  receipt_id?: string
  product_id: string
  product_name?: string
  product_code?: string
  qty: number
  uom: string
  unit_price: number
  tax_rate: number
  discount_pct: number
  line_total: number
}

export interface POSReceipt {
  id?: string
  tenant_id?: string
  register_id: string
  shift_id: string
  cashier_id?: string | null
  cashier_name?: string | null
  number?: string
  status: 'draft' | 'paid' | 'voided' | 'invoiced' | 'unpaid'
  customer_id?: string
  invoice_id?: string
  gross_total: number
  tax_total: number
  subtotal?: number
  currency: string
  paid_at?: string
  created_at?: string
  lines: POSReceiptLine[]
  payments?: POSPayment[]
}

export interface POSPayment {
  id?: string
  receipt_id: string
  method: 'cash' | 'card' | 'store_credit' | 'link'
  amount: number
  ref?: string
  paid_at?: string
}

export interface StoreCredit {
  id: string
  tenant_id: string
  code: string
  customer_id?: string
  currency: string
  amount_initial: number
  amount_remaining: number
  expires_at?: string
  status: 'active' | 'redeemed' | 'expired' | 'void'
  created_at: string
}

export interface Product {
  id: string
  code?: string
  sku?: string
  name: string
  price: number
  tax_rate: number
  stock_qty?: number
  stock?: number
  image_url?: string
  category?: string
  // POS extras
  uom?: string
  weight_required?: boolean
}

export interface ShiftOpenRequest {
  register_id: string
  opening_float: number
  cashier_id?: string
}

export interface ShiftCloseRequest {
  shift_id: string
  closing_cash: number
  loss_amount?: number
  loss_note?: string
}

export interface ShiftSummary {
  pending_receipts: number
  items_sold: Array<{
    product_id: string | null
    name: string
    code: string
    qty_sold: number
    subtotal: number
    stock: Array<{
      warehouse_id: string
      warehouse_name: string
      qty: number
    }>
  }>
  sales_total: number
  receipts_count: number
  payments?: Record<string, number>
}

export interface ReceiptCreateRequest {
  register_id: string
  shift_id: string
  cashier_id?: string
  customer_id?: string
  currency?: string
  lines: POSReceiptLine[]
  payment_method?: string
}

export interface ReceiptToInvoiceRequest {
  customer: {
    name: string
    tax_id: string
    country: string
    address?: string
    email?: string
  }
  series?: string
}

export interface RefundRequest {
  reason: string
  refund_method: 'original' | 'cash' | 'store_credit'
  line_ids?: string[]
  restock?: boolean
  store_credit_expiry_months?: number
}

export interface PaymentLinkRequest {
  // Either invoice_id or receipt_id may be provided depending on flow
  invoice_id?: string
  receipt_id?: string
  provider?: 'stripe' | 'kushki' | 'payphone'
  amount?: number
  currency?: string
  description?: string
  metadata?: Record<string, any>
  success_url?: string
  cancel_url?: string
}

export interface CartItem extends POSReceiptLine {
  product?: Product
}

export interface DocSeries {
  id: string
  tenant_id: string
  register_id?: string
  doc_type: 'R' | 'F' | 'C'
  name: string
  current_no: number
  reset_policy: 'yearly' | 'never'
  active: boolean
  created_at: string
}

export interface EInvoiceStatus {
  kind: 'SRI' | 'SII'
  ref: string
  status: 'pending' | 'authorized' | 'rejected'
  clave_acceso?: string
  error_message?: string
  submitted_at?: string
  xml_content?: string
}

export interface PaymentLink {
  id: string
  invoice_id: string
  invoice_number: string
  provider: 'stripe' | 'kushki' | 'payphone'
  url: string
  amount: number
  currency: string
  status: 'pending' | 'completed' | 'expired'
  created_at: string
  expires_at?: string
}

export interface Transaction {
  id: string
  invoice_id: string
  invoice_number: string
  provider: string
  amount: number
  currency: string
  status: 'pending' | 'success' | 'failed'
  ref: string
  created_at: string
}

File: apps/backend/app/schemas/pos.py
"""Schemas Pydantic para POS"""

from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ==================== REGISTER SCHEMAS ====================
class POSRegisterBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    store_id: UUID | None = None
    active: bool = True


class POSRegisterCreate(POSRegisterBase):
    pass


class POSRegisterUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=100)
    active: bool | None = None


class POSRegisterResponse(POSRegisterBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    tenant_id: UUID
    created_at: datetime


# ==================== SHIFT SCHEMAS ====================
class POSShiftBase(BaseModel):
    opening_float: Decimal = Field(..., ge=0, decimal_places=2)


class POSShiftCreate(POSShiftBase):
    register_id: UUID


class POSShiftClose(BaseModel):
    closing_total: Decimal = Field(..., ge=0, decimal_places=2)


class POSShiftResponse(POSShiftBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    register_id: UUID
    opened_by: UUID
    opened_at: datetime
    closed_at: datetime | None = None
    closing_total: Decimal | None = None
    status: str


# ==================== RECEIPT LINE SCHEMAS ====================
class POSReceiptLineBase(BaseModel):
    product_id: UUID
    qty: Decimal = Field(..., gt=0, decimal_places=3)
    uom: str = Field(default="unit", max_length=20)
    unit_price: Decimal = Field(..., ge=0, decimal_places=4)
    tax_rate: Decimal = Field(..., ge=0, le=1, decimal_places=4)
    discount_pct: Decimal = Field(default=Decimal("0"), ge=0, le=100, decimal_places=2)


class POSReceiptLineCreate(POSReceiptLineBase):
    pass


class POSReceiptLineResponse(POSReceiptLineBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    receipt_id: UUID
    line_total: Decimal


# ==================== PAYMENT SCHEMAS ====================
class POSPaymentBase(BaseModel):
    method: str = Field(..., pattern="^(cash|card|store_credit|link)$")
    amount: Decimal = Field(..., gt=0, decimal_places=2)
    ref: str | None = None


class POSPaymentCreate(POSPaymentBase):
    pass


class POSPaymentResponse(POSPaymentBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    receipt_id: UUID
    paid_at: datetime


# ==================== RECEIPT SCHEMAS ====================
class POSReceiptBase(BaseModel):
    customer_id: UUID | None = None
    currency: str = Field(default="EUR", pattern="^(EUR|USD)$")


class POSReceiptCreate(POSReceiptBase):
    register_id: UUID
    shift_id: UUID
    lines: list[POSReceiptLineCreate] = Field(..., min_length=1)
    payments: list[POSPaymentCreate] | None = None


class POSReceiptUpdate(BaseModel):
    status: str | None = Field(None, pattern="^(draft|paid|voided|invoiced)$")
    customer_id: UUID | None = None


class POSReceiptResponse(POSReceiptBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    tenant_id: UUID
    register_id: UUID
    shift_id: UUID
    cashier_id: UUID | None = None
    number: str
    status: str
    invoice_id: UUID | None = None
    gross_total: Decimal
    tax_total: Decimal
    paid_at: datetime | None = None
    created_at: datetime
    lines: list[POSReceiptLineResponse] = []
    payments: list[POSPaymentResponse] = []


# ==================== INVOICE CONVERSION SCHEMAS ====================
class CustomerDataForInvoice(BaseModel):
    name: str = Field(..., min_length=1)
    tax_id: str = Field(..., min_length=1)
    country: str = Field(..., pattern="^(ES|EC)$")
    address: str | None = None
    email: str | None = None


class ReceiptToInvoiceRequest(BaseModel):
    customer: CustomerDataForInvoice
    series: str | None = None


# ==================== REFUND SCHEMAS ====================
class ReceiptRefundRequest(BaseModel):
    reason: str = Field(..., min_length=1, max_length=500)
    refund_method: str = Field(..., pattern="^(cash|card|store_credit)$")
    store_credit_months: int | None = Field(default=12, ge=1, le=24)
